File: pa1/tli.hs
-- Daniel Xiong dxiong5@ucsc.edu id#1660652
-- tli.hs
-- due 10/21/19
-- Pair Programming partner: Scott Zin nzin@ucsc.edu id#1679510


import Data.Char
import System.IO
import System.Environment
import System.Exit
import Data.Typeable

-- maps labels line numbers and variables to values - uses float for line numbers for simplicity
type SymTable = [(String,Float)]

data Expr = 
     Constant Float |
     Var String |
     Plus Expr Expr |
     Minus Expr Expr | 
     Times Expr Expr | 
     Div Expr Expr |
     LT_ Expr Expr |
     GT_ Expr Expr |
     LE_ Expr Expr |
     GE_ Expr Expr |
     EQ_ Expr Expr |
     NEQ_ Expr Expr |
     ExprError String deriving (Show, Eq) 

data Stmt =
     Let String Expr |
     Print [Expr] |
     If Expr String |
     Input String deriving (Show) 

-- dummy predicate that is supposed to check if a string is a label which is a string ending with ":"
isLabel :: String -> Bool
isLabel str = if ((last str) == ':') then True else False

-- takes a list of tokens as strings and returns the parsed expression
parseExpr :: [String] -> Expr
parseExpr [x] = if (isAlpha (head x)) then (Var x) else if (isPunctuation (head x)) then (ExprError x) else (Constant (read x))-- pattern matches the most basic elements, ie: var names and constants (turns constant into int)
parseExpr (e1:"+":e2:[]) = Plus (parseExpr [e1]) (parseExpr [e2]) -- parses addition expression ie: ["x", "+", "1"] = Plus x 1
parseExpr (e1:"-":e2:[]) = Minus (parseExpr [e1]) (parseExpr [e2])
parseExpr (e1:"*":e2:[]) = Times (parseExpr [e1]) (parseExpr [e2])
parseExpr (e1:"/":e2:[]) = Div (parseExpr [e1]) (parseExpr [e2])
parseExpr (e1:"<":e2:[]) = LT_ (parseExpr [e1]) (parseExpr[e2])
parseExpr (e1:">":e2:[]) = GT_ (parseExpr [e1]) (parseExpr[e2])
parseExpr (e1:"<=":e2:[]) = LE_ (parseExpr [e1]) (parseExpr[e2])
parseExpr (e1:">=":e2:[]) = GE_ (parseExpr [e1]) (parseExpr[e2])
parseExpr (e1:"==":e2:[]) = EQ_ (parseExpr [e1]) (parseExpr [e2])
parseExpr (e1:"!=":e2:[]) = NEQ_ (parseExpr [e1]) (parseExpr [e2])


-- helper function for parseStmt
-- takes in unprocessed [String] and returns [Expr]
parseStmtPrintHelper :: [String] -> [String] -> [Expr] -> [Expr]
parseStmtPrintHelper [x] curString exprList = if ((last x) /= '"') then exprList++[(parseExpr (curString++[x]))] else exprList++[(parseExpr [(unwords (curString++[x]))])]
parseStmtPrintHelper (x:xs) curString exprList = 
    if (x == ",")
    then if ((last (last curString)) == '"') then parseStmtPrintHelper xs [] (exprList++[(parseExpr [(unwords curString)])]) else parseStmtPrintHelper xs [] (exprList++[(parseExpr curString)])
    else parseStmtPrintHelper xs (curString++[x]) exprList

-- takes the first token which should be a keyword and a list of the remaining tokens and returns the parsed Stmt
parseStmt :: String -> [String] -> Float -> Stmt
parseStmt "let" (v:"=":expr) _ = Let v (parseExpr expr)
parseStmt "print" expr _ = Print (parseStmtPrintHelper expr [] [])
parseStmt "if" rest _ = If (parseExpr (init (init rest))) (last rest)
parseStmt "input" [varName] _ = Input varName
parseStmt _ _ lineNum = error ("Syntax error on line "++(show lineNum))


-- recursive function that calls parseLine on each iteration and builds a list of Statements
runParseLine :: [[String]] -> [Stmt] -> SymTable -> Float -> ([Stmt], SymTable)
runParseLine [] stmtList env _ = (stmtList, env)
runParseLine (head:rest) stmtList env lineNumber = 
    let (statement, env1) = parseLine head env lineNumber in runParseLine rest (stmtList++[statement]) env1 (lineNumber+1)

-- takes a list of tokens and returns the parsed statement - the statement may include a leading label
parseLine :: [String] -> SymTable -> Float -> (Stmt, SymTable)
parseLine (first:rest) env lineNum =
     if (isLabel first) 
     then parseLine rest ((first,lineNum):env) lineNum -- if the first string is a label, then run parseLine again on the remaining strings, which now should be a Statement
     else ((parseStmt first rest lineNum), env) -- if the first string isnt a label, then this line is a Statement

-- takes a variable name and a ST and returns the value of that variable or zero if the variable is not in the ST
lookupVar :: String -> SymTable -> Float -> Bool -> Float
lookupVar name [] lineNum isLab = if isLab then error ("Illegal goto "++(show (init name))++" at line "++(show lineNum)) else error ("Undefined variable name "++(show name)++" at line "++(show lineNum))
lookupVar name ((id,v):rest) lineNum isLab = if (id == name) then v else lookupVar name rest lineNum isLab


-- evaluates the given Expr with the variable values found in the given ST
eval :: Expr -> SymTable -> Float -> Float
eval (Var v) env lineNum = lookupVar v env lineNum False
eval (Constant v) _ _ = v
eval (Plus e1 e2) env lineNum = (eval e1 env lineNum) + (eval e2 env lineNum)
eval (Minus e1 e2) env lineNum = (eval e1 env lineNum) - (eval e2 env lineNum)
eval (Times e1 e2) env lineNum = (eval e1 env lineNum) * (eval e2 env lineNum)
eval (Div e1 e2) env lineNum = (eval e1 env lineNum) / (eval e2 env lineNum)
eval (LT_ e1 e2) env lineNum = if (eval e1 env lineNum) < (eval e2 env lineNum) then 1 else 0
eval (GT_ e1 e2) env lineNum = if (eval e1 env lineNum) > (eval e2 env lineNum) then 1 else 0
eval (LE_ e1 e2) env lineNum = if (eval e1 env lineNum) <= (eval e2 env lineNum) then 1 else 0
eval (GE_ e1 e2) env lineNum = if (eval e1 env lineNum) >= (eval e2 env lineNum) then 1 else 0
eval (EQ_ e1 e2) env lineNum = if (eval e1 env lineNum) == (eval e2 env lineNum) then 1 else 0
eval (NEQ_ e1 e2) env lineNum = if (eval e1 env lineNum) /= (eval e2 env lineNum) then 1 else 0

-- helper function for perform's Print pattern match
-- takes list of expressions [Expr] and adds them to output
printHelper :: [Expr] -> SymTable -> String -> Float -> String
printHelper [] _ output lineNum = output++"\n"
printHelper ((ExprError e):es) env output lineNum = 
    let out = unwords.words $ reverse . dropWhile (not.isAlpha) . reverse $ dropWhile (not.isAlpha) (show e) 
    in printHelper es env (output++out++" ") lineNum
printHelper (e:es) env output lineNum = let out = show (eval e env lineNum) in printHelper es env (output++out++" ") lineNum

-- given a statement, a ST, line number, input and previous output, return an updated ST, input, output, and line number
-- this starter version ignores the input and line number
-- Stmt, SymTable, progCounter, input, output, (SymTable', input', output', progCounter)
-- Stmt : Print [(x + 1), (x + 2)]
perform :: Stmt -> SymTable -> Float -> [String] -> String -> (SymTable, [String], String, Float)
perform (Print exprList) env lineNum input output = (env, input, (printHelper exprList env output lineNum), lineNum+1) -- Print e now takes e as [String], so find a way to make it print the list
perform (Let id e) env lineNum input output = ((id,(eval e env lineNum)):env, input, output, lineNum+1)
perform (If expr label) env lineNum input output = if ((eval expr env lineNum) == 1) then (env, input, output, (lookupVar (label++":") env lineNum True)) else (env, input, output, lineNum+1)
perform (Input varName) env lineNum [] output = error ("Prelude.read: no parse")
perform (Input varName) env lineNum (x:xs) output = ((varName,(read x)):env, xs, output, lineNum+1)

-- helper function for run 
-- returns Stmt at its respective line number
getStmtAtLineNum :: [Stmt] -> Float -> Stmt
getStmtAtLineNum (x:_) 1 = x
getStmtAtLineNum (x:xs) i = getStmtAtLineNum xs (i-1)

-- given a list of Stmts, a ST, and current output, perform all of the statements in the list and return the updated output String
run :: [Stmt] -> SymTable -> [String] -> String -> Float -> String
run stmtList env input output lineNum = 
    if (lineNum == ((fromIntegral (length stmtList)) + 1.0))
    then output
    else let (env1, input1, output1, newLineNum) = perform (getStmtAtLineNum stmtList lineNum) env lineNum input output in run stmtList env1 input1 output1 newLineNum

-- given list of list of tokens, a ST, return the list of parsed Stmts and ST storing mapping of labels to line numbers
parseTest :: [[String]] -> SymTable -> ([Stmt], SymTable)
parseTest []  st = ([], st)
parseTest allLines env = runParseLine allLines [] env 1


main = do
     input <- getContents -- reads user input from stdin
     args <- getArgs -- gets command line args
     pfile <- openFile (head args) ReadMode -- opens file
     contents <- hGetContents pfile -- gets file contents
     let input1 = words input -- formats user input into list
     let allLines = (map words (lines contents)) -- formats file contents into a [[String]]
     let (stmtList, env) = runParseLine allLines [] [] 1 -- gets list of statements and symtable currently storing only labels
     let output = run stmtList env input1 "" 1 -- gets output from run
     putStr output
     hClose pfile

File: pa3/tli.py
#! /usr/bin/env python3
'''
Daniel Xiong dxiong5@ucsc.edu id#1660652
tli.py
Due 11/22/2019
'''
import fileinput
import io
import sys

# used to store a parsed TL expressions which are
# constant numbers, constant strings, variable names, and binary expressions
class Expr:
	def __init__(self,operator,op1=None,op2=None):
		"""
		operator types:
		var, constant, printstring, 
		plus, minus, mult, div, lt, gt, le, ge, eq, neq
		"""
		
		self.op1 = op1
		self.operator = operator
		self.op2 = op2

	def __str__(self):
		if self.op2 == None:
			return self.operator + " " + str(self.op1)
		else:
			return str(self.op1) + " " + self.operator + " " +  str(self.op2)

	def eval(self, symTable, lineNum):
		""" evaluates Expr objects """

		if self.operator == "var":
			if self.op1 in symTable:
				return symTable[self.op1]
			else:
				print("Undefined variable {} at line {}.".format(self.op1, lineNum))
				sys.exit()
		elif self.operator == "constant":
			return self.op1
		elif self.operator == "printstring":
			return self.op1
		elif self.operator == "plus":
			return self.op1.eval(symTable, lineNum)+self.op2.eval(symTable, lineNum)
		elif self.operator == "minus":
			return self.op1.eval(symTable, lineNum)-self.op2.eval(symTable, lineNum)
		elif self.operator == "mult":
			return self.op1.eval(symTable, lineNum)*self.op2.eval(symTable, lineNum)
		elif self.operator == "div":
			return self.op1.eval(symTable, lineNum)/self.op2.eval(symTable, lineNum)
		elif self.operator == "lt":
			if self.op1.eval(symTable, lineNum) < self.op2.eval(symTable, lineNum):
				return 1
			else:
				return 0
		elif self.operator == "gt":
			if self.op1.eval(symTable, lineNum) > self.op2.eval(symTable, lineNum):
				return 1
			else:
				return 0
		elif self.operator == "le":
			if self.op1.eval(symTable, lineNum) <= self.op2.eval(symTable, lineNum):
				return 1
			else:
				return 0
		elif self.operator == "ge":
			if self.op1.eval(symTable, lineNum) >= self.op2.eval(symTable, lineNum):
				return 1
			else:
				return 0
		elif self.operator == "eq":
			if self.op1.eval(symTable, lineNum) == self.op2.eval(symTable, lineNum):
				return 1
			else:
				return 0
		elif self.operator == "neq":
			if self.op1.eval(symTable, lineNum) != self.op2.eval(symTable, lineNum):
				return 1
			else:
				return 0
		else:
			print("Syntax error on line {}.".format(lineNum))
			sys.exit()

class Scanner:
    def __init__(self):
        self.currentLine = []
    def readInt(self):
        if len(self.currentLine) == 0:
            self.currentLine = input().split()
        next = self.currentLine[0]
        self.currentLine = self.currentLine[1:]
        return next

class Stmt :
	""" Stmt object """

	def __init__(self,keyword,var=None,gotoLabel=None,exprs=None):
		self.keyword = keyword
		self.exprs = exprs
		self.var = var
		self.gotoLabel = gotoLabel

	def __str__(self):
		others = ""
		if self.keyword == 'let':
			others = others + " " + self.var
			others = others + " " + str(self.exprs)
		elif self.keyword == 'print':
			for exp in self.exprs:
				others = others + " " + str(exp)
		elif self.keyword == 'if':
			others = others + " " + str(self.exprs)
			others = others + ' goto ' + self.gotoLabel
		elif self.keyword == 'input':
			others = others + " " + self.var
		return self.keyword + others

	def perform(self, output, symTable, lineNum, scan):
		""" performs a Stmt object, returns updated parameters """

		if self.keyword == 'let': # if let statement, update symtable and lineNum and return
			symTable[self.var] = self.exprs.eval(symTable, lineNum)
			return (output, symTable, lineNum+1)
		elif self.keyword == 'print': # if print statement, update output and lineNum and return
			for e in self.exprs:
				if e == self.exprs[-1]:
					output = output + str(e.eval(symTable, lineNum)) + '\n'
				else:
					output = output + str(e.eval(symTable, lineNum)) + ' '
			return (output, symTable, lineNum+1)
		elif self.keyword == 'if': # if ifgoto statement, check if label is in symtable and update lineNum and return
			if self.exprs.eval(symTable, lineNum) != 0:
				if self.gotoLabel in symTable:
					return (output, symTable, symTable[self.gotoLabel])
				else:
					print("Illegal goto {} at line {}.".format(self.gotoLabel[:-1], lineNum))
					sys.exit()
			else:
				return (output, symTable, lineNum+1)
		elif self.keyword == 'input': # if input statement, update symtable and return
			try:
				i = float(scan.readInt())
			except:
				print('Illegal or missing input.')
				sys.exit()
			symTable[self.var] = i

			return (output, symTable, lineNum+1)
		elif self.keyword == 'no-op':
			return (output, symTable, lineNum+1)


def parseLine(lines, stmtList, symTable):
	""" sends each line to parseStmt, returns list of statements and updated symtable """

	if len(lines) == 0:
		return (stmtList, symTable)

	for i, line in enumerate(lines):
		if len(line) == 0:
			stmtList.append(Stmt(keyword="no-op"))
			continue
		lineNum = i + 1
		if line[0].endswith(':'): # if the line starts with a label, remove label and run parseStmt on rest of the line
			symTable[line[0]] = lineNum
			statement = parseStmt(line[1:], lineNum)
		else:
			statement = parseStmt(line, lineNum)
		stmtList.append(statement)

	return (stmtList, symTable)

def parseStmtPrintHelper(tokens, lineNum):
	""" helper function for parseStmt, handles print statement """

	curExpr = []
	exprList = []
	while len(tokens) != 0:
		s = tokens.pop(0)
		
		if len(tokens) == 0: # if only one thing left in tokens list
			curExpr.append(s)
			if s.endswith('"'):
				string = ' '.join(curExpr)
				exprList.append(parseExpr(list([string]), lineNum))
			else:
				exprList.append(parseExpr(curExpr, lineNum))
		
		if s == ',': # if current token is the end of an expr, then parse the expr and reset the curExpr list
			if curExpr[-1].endswith('"'):
				string = ' '.join(curExpr)
				exprList.append(parseExpr(list([string]), lineNum))
			else:
				exprList.append(parseExpr(curExpr, lineNum))
			curExpr = []
		elif s.endswith(','): # if current token is the end of an expr, then parse the expr and reset the curExpr list
			curExpr.append(s)
			if s[-2] == '"':
				string = ' '.join(curExpr).strip(',')
				exprList.append(parseExpr(list([string]), lineNum))
				curExpr = []
			elif curExpr[0][0] != '"':
				string = ' '.join(curExpr).strip(',')
				curExpr = string.split()
				exprList.append(parseExpr(curExpr, lineNum))
				curExpr = []
		else:
			curExpr.append(s)

	return exprList

def parseStmt(line, lineNum):
	""" identifies statement type and returns Stmt object """ 

	if line[0] == 'let' and line[2] == '=':
		e = parseExpr(line[3:], lineNum)
		return Stmt(keyword=line[0], var=line[1], exprs=e)
	elif line[0] == 'print':
		exprList = parseStmtPrintHelper(line[1:], lineNum)
		return Stmt(keyword=line[0], exprs=exprList)
	elif line[0] == 'if' and line[-2] == 'goto':
		e = parseExpr(line[1:-2], lineNum)
		l = line[-1] + ':'
		return Stmt(keyword=line[0], gotoLabel=l, exprs=e)
	elif line[0] == 'input':
		return Stmt(keyword=line[0], var=line[1])
	else:
		print("Syntax error on line {}.".format(lineNum))
		sys.exit()

def parseExpr(tokens, lineNum):
	""" parses each expression and returns Expr object """

	# if there is only one token
	if len(tokens) == 1: 
		x = tokens[0]
		if x[0].isalpha(): # if the token is a variable
			return Expr(op1=x, operator='var')
		elif x.isnumeric(): # if the token is a number
			return Expr(op1=float(x), operator='constant')
		else: # token must be a print string
			return Expr(op1=x.strip('"'), operator='printstring')

	# if actual expression
	if tokens[1] == '+':
		e1 = parseExpr(tokens[:1], lineNum)
		e2 = parseExpr(tokens[2:], lineNum)
		return Expr(op1=e1, operator='plus',op2=e2)
	elif tokens[1] == '-':
		e1 = parseExpr(tokens[:1], lineNum)
		e2 = parseExpr(tokens[2:], lineNum)
		return Expr(op1=e1, operator='minus',op2=e2)
	elif tokens[1] == '*':
		e1 = parseExpr(tokens[:1], lineNum)
		e2 = parseExpr(tokens[2:], lineNum)
		return Expr(op1=e1, operator='mult',op2=e2)
	elif tokens[1] == '/':
		e1 = parseExpr(tokens[:1], lineNum)
		e2 = parseExpr(tokens[2:], lineNum)
		return Expr(op1=e1, operator='div',op2=e2)
	elif tokens[1] == '<':
		e1 = parseExpr(tokens[:1], lineNum)
		e2 = parseExpr(tokens[2:], lineNum)
		return Expr(op1=e1, operator='lt',op2=e2)
	elif tokens[1] == '>':
		e1 = parseExpr(tokens[:1], lineNum)
		e2 = parseExpr(tokens[2:], lineNum)
		return Expr(op1=e1, operator='gt',op2=e2)
	elif tokens[1] == '<=':
		e1 = parseExpr(tokens[:1], lineNum)
		e2 = parseExpr(tokens[2:], lineNum)
		return Expr(op1=e1, operator='le',op2=e2)
	elif tokens[1] == '>=':
		e1 = parseExpr(tokens[:1], lineNum)
		e2 = parseExpr(tokens[2:], lineNum)
		return Expr(op1=e1, operator='ge',op2=e2)
	elif tokens[1] == '==':
		e1 = parseExpr(tokens[:1], lineNum)
		e2 = parseExpr(tokens[2:], lineNum)
		return Expr(op1=e1, operator='eq',op2=e2)
	elif tokens[1] == '!=':
		e1 = parseExpr(tokens[:1], lineNum)
		e2 = parseExpr(tokens[2:], lineNum)
		return Expr(op1=e1, operator='neq',op2=e2)
	else: 
		print("Syntax error on line {}.".format(lineNum))
		sys.exit()

if __name__ == '__main__':

	# reads file
	infile = sys.argv[1]
	with open(infile, "r+") as fin:
		# contents is a 2D list, where each row is a line, with line number zero indexed
		contents = []
		for line in fin.readlines():
			tokens = line.strip('\n').strip('\t').strip('\r').split()
			contents.append(tokens)

	# gets parsed list of statements and symtable with label-linenum pairs
	stmt_list, symTable = parseLine(contents, [], {})

	scan = Scanner()
	output = ''
	lineNum = 1
	stmt_list.insert(0,None) # 1 indexing
	while lineNum < len(stmt_list):
		# executes each statement
		output, symTable, lineNum = stmt_list[lineNum].perform(output=output, symTable=symTable, lineNum=lineNum, scan=scan)

	# prints final output
	print(output, end='')